### PR TITLE
i1854 additional mempool endpoints

### PR DESCRIPTION
**Pull request by pragmaxim**
Closes #https://github.com/ergoplatform/ergo/issues/1854

I added the new endpoints under existing `/transactions/unconfirmed/` endpoint as it is all searching through unconfirmed transactions ... probably better than involving the new word `mempool`.

@MrStahlfelge I implemented it all besides :
```
POST: mempool/transactions/inputs/?check={true|false}
POST: mempool/transactions/outputs/?check={true|false}
```
As they seem hard to reason about, the description : 
>> Batch get a list of inputs filtered by multiple boxIds. If check=true it must check if boxIds list elements are present on mempool as inputs and only return a list of boxId of included ones.

says : "Either return Boxes Or only BoxIds if `check=true` and they are present in mempool. I'd be happy to implement something like this but imho it should be easier to reason about.

### CODE REVIEW

File: src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala [hunk 1]
```diff
@@ -87,4 +124,120 @@ case class TransactionsApiRoute(readersHolder: ActorRef,
...
+
+  /** Get unconfirmed transaction by its id */
+  def getUnconfirmedTxByIdR: Route =
+    (pathPrefix("unconfirmed" / "byTransactionId") & get & modifierId) { modifierId =>
+      ApiResponse(getMemPool.map(_.modifierById(modifierId)))
+    }
+
+  /** Collect all transactions which inputs or outputs contain given ErgoTree hex */
+  def getUnconfirmedTxByErgoTreeR: Route =
+    (pathPrefix("unconfirmed" / "byErgoTree") & post & entity(as[Json])) { body =>
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala:147]: I think paging should be implemented for this query, for some trees there could be thousands of transactions

File: src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala [hunk 2]
```diff
@@ -87,4 +124,120 @@ case class TransactionsApiRoute(readersHolder: ActorRef,
...
+  def getUnconfirmedOutputByBoxIdR: Route =
+    (pathPrefix("unconfirmed" / "outputs" / "byBoxId") & get & boxId) { boxId =>
+      ApiResponse(
+        getMemPool.map(_.getAll.flatMap(_.transaction.outputs.filter(_.id.sameElements(boxId))).headOption)
+      )
+    }
+
+  /** Collect all tx outputs which contain given ErgoTree hex */
+  def getUnconfirmedOutputByErgoTreeR: Route =
+    (pathPrefix("unconfirmed" / "outputs" / "byErgoTree") & post & entity(as[Json])) { body =>
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala:205]: I think paging should be implemented here as well

File: src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala [hunk 3]
```diff
@@ -87,4 +124,120 @@ case class TransactionsApiRoute(readersHolder: ActorRef,
...
+  def getUnconfirmedOutputByTokenIdR: Route =
+    (pathPrefix("unconfirmed" / "outputs" / "byTokenId") & get & tokenId) { tokenId =>
+      ApiResponse(
+        getMemPool.map(_.getAll.flatMap(_.transaction.outputs.filter(_.additionalTokens.exists(_._1.sameElements(tokenId)))))
+      )
+    }
+
+  /** Collect all tx outputs which contain all given Registers */
+  def getUnconfirmedOutputByRegistersR: Route =
+    (pathPrefix("unconfirmed" / "outputs" / "byRegisters") & post & entity(as[Json])) { body =>
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala:226]: I think paging should be implemented here as well

File: src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala [hunk 4]
```diff
@@ -87,4 +124,121 @@ case class TransactionsApiRoute(readersHolder: ActorRef,
...
+  /** Collect all transactions which inputs or outputs contain given ErgoTree hex */
+  def getUnconfirmedTxsByErgoTreeR: Route =
+    (pathPrefix("unconfirmed" / "byErgoTree") & post & entity(as[Json]) & txPaging) { case (body, offset, limit) =>
+      body.as[String] match {
+        case Left(ex) =>
+          ApiError(StatusCodes.BadRequest, ex.getMessage())
+        case Right(ergoTree) =>
+          ApiResponse(
+            getMemPool.flatMap { pool =>
+              val allTxs = pool.getAll.slice(offset, offset + limit)
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala:154]: So if I am asking for 50 transactions, your code may return 5, as slicing happens before filtering. The same is true for all other new methods with paging

**Review by kushti — CHANGES_REQUESTED** [review]

File: src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala (hunk 4 above)
Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/http/api/TransactionsApiRoute.scala]: True, slicing result should happen after filtering

**Review by kushti — APPROVED** [review]

**Pull request merged**